### PR TITLE
diskWithAllParts always fallsback to blake2b for bitrot check

### PR DESCRIPTION
**Pull request by dlind1974**
For erasure coded Minio the disksWithAllParts in the XL layer always passes the erasure coding algorithm name when constructing a new hash. This results in that function newHash will fallback to blake2b hash algorithm.

## Description
The disksWithAllParts function has been updated to use correct algorithm which is stored in the Checksum part of the parts meta data. Previously the erasure coding algorithm stored directly in the root of the parts meta data was passed.

## Motivation and Context
If an alternative hash algorithm like HashSha256 is used it will cause a mismatch since the checksum algorithm name is not passed when creating the hashSum instance.

## How Has This Been Tested?
Manually testing with debug printing
New unit test has been written to catch the bug.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/xl-v1-healing-common_test.go [hunk 1]
```diff
@@ -330,3 +330,110 @@ func TestListOnlineDisks(t *testing.T) {
...
+		if _, ok := diskFailures[diskIndex]; ok {
+			if disk != nil {
+				t.Errorf("Disk not filtered as expected, disk: %d", diskIndex)
+			}
+			if errs[diskIndex] == nil {
+				t.Errorf("Expected error not received, diskIndex: %d", diskIndex)
+			}
+		} else {
+			if disk == nil {
+				t.Errorf("Disk erronous filtered, disk: %d", diskIndex)
```
Comment (krisis) [inline cmd/xl-v1-healing-common_test.go]: minor typo: s/erronous/erroneous

File: cmd/xl-v1-healing-common_test.go [hunk 2]
```diff
@@ -330,3 +330,110 @@ func TestListOnlineDisks(t *testing.T) {
...
+		t.Errorf("Unexpected number of disks: %d", len(filteredDisks))
+	}
+
+	for diskIndex, disk := range filteredDisks {
+		if errs[diskIndex] != nil {
+			t.Errorf("Unexpected error %s", errs[diskIndex])
+		}
+
+		if disk == nil {
+			t.Errorf("Disk erronous filtered, disk: %d", diskIndex)
```
Comment (krisis) [inline cmd/xl-v1-healing-common_test.go]: minor typo: s/erronous/erroneous

File: cmd/xl-v1-healing-common_test.go [hunk 3]
```diff
@@ -330,3 +330,110 @@ func TestListOnlineDisks(t *testing.T) {
...
+
+	partsMetadata, errs := readAllXLMetadata(xlDisks, bucket, object)
+	if err != nil {
+		t.Fatalf("Failed to read xl meta data %v", err)
+	}
+
+	// Replace the default blake2b erasure algorithm to HashSha256 and test that
+	// disks are excluded
+	diskFailures := make(map[int]int)
+	// key = disk index, value = part index with hash mistmatch
```
Comment (krisis) [inline cmd/xl-v1-healing-common_test.go]: minor typo: s/mistmatch/mismatch

**Review by krisis — COMMENTED** [review]
@dlind1974 Thank you for the fix. The changes look good to me, except for a few minor typos. I will test it once locally before I approve the PR.

**Comment by harshavardhana** [comment]
Thanks a lot @dlind1974 - great catch and nice fix.  Thanks for adding the tests as well.

**Review by vadmeste — APPROVED** [review]
LGTM & tested. Thanks @dlind1974

**Comment by dlind1974** [comment]
I have made a force push to the bugfix-xl-disk-parts-hash-algo branch on dlind1974/minio with updates for the typos found by @krisis.

**Comment by dlind1974** [comment]
I made a force push with fixes for typos.

**Comment by harshavardhana** [comment]
> I made a force push with fixes for typos.

That is expected - thank you @dlind1974

**Review by krisis — APPROVED** [review]
LGTM.

**Comment by dlind1974** [comment]
I don know if everything is ok, but I think that the latest approval in the review got invalidated due to my updates.

**Comment by harshavardhana** [comment]
@dlind1974 everything is fine no need to worry.

File: cmd/xl-v1-healing-common.go [hunk 4]
```diff
@@ -252,39 +252,39 @@ func xlHealStat(xl xlObjects, partsMetadata []xlMetaV1, errs []error) HealObject
...
 		}
 		// disk has a valid xl.json but may not have all the
 		// parts. This is considered an outdated disk, since
 		// it needs healing too.
-		for _, part := range partsMetadata[index].Parts {
+		for partIndex, part := range partsMetadata[diskIndex].Parts {
 			// compute blake2b sum of part.
 			partPath := filepath.Join(object, part.Name)
-			hash := newHash(partsMetadata[index].Erasure.Algorithm)
+			hash := newHash(partsMetadata[diskIndex].Erasure.Checksum[partIndex].Algorithm)
```
Comment (vadmeste) [inline cmd/xl-v1-healing-common.go]: Actually I found a little problem here when I reviewed again, the order of elements in Parts and Checksum is not *always* the same. so CheckSum[partIndex] and Parts[partIndex] are not always corresponding to the same part. I'll try to suggest a simple solution for this.

Comment (vadmeste) [reply cmd/xl-v1-healing-common.go]: For example, if I enforce mc to upload parts with a reverse order, I get this in xl.json:

```
{
  "version": "1.0.1",
  "format": "xl",
  "stat": {
    "size": 134217728,
    "modTime": "2017-06-10T11:23:46.742721655Z"
  },
  "erasure": {
    "algorithm": "klauspost/reedsolomon/vandermonde",
    "data": 2,
    "parity": 2,
    "blockSize": 10485760,
    "index": 1,
    "distribution": [
      1,
      2,
      3,
      4
    ],
    "checksum": [
      {
        "name": "part.2",
        "algorithm": "blake2b",
        "hash": "76a510f220d2ce6423a92b2df2b0a3fe5ce0c77a575a4ead76dca3e80bcc56c79e5a245c7ca0d7add2e73aebf43abf6e9a009d916799fba1922fe6183498356b"
      },
      {
        "name": "part.1",
        "algorithm": "blake2b",
        "hash": "dfec7a954c6c9c616d586751960682dab9cd2e825f6d3aab97aec523f50c092245d05600762235790b45919531b4516bb56db207f5afb2ec7395254f532f4b8a"
      }
    ]
  },
  "minio": {
    "release": "DEVELOPMENT.GOGET"
  },
  "meta": {
    "content-type": "application/octet-stream",
    "etag": "51b1206b454b32ac37305a2b544b0511-2"
  },
  "parts": [
    {
      "number": 1,
      "name": "part.1",
      "etag": "48255d8705e103628f6b443fb70cb533",
      "size": 67108864
    },
    {
      "number": 2,
      "name": "part.2",
      "etag": "8e018c99de4371f5f7634f74aff04683",
      "size": 67108864
    }
  ]
}
```

Comment (vadmeste) [reply cmd/xl-v1-healing-common.go]: It looks like replacing `newHash(partsMetadata[diskIndex].Erasure.Checksum[partIndex].Algorithm)` with
`newHash(partsMetadata[diskIndex].Erasure.GetCheckSumInfo(part.Name).Algorithm)` is enough to fix the problem. Can you check if my suggestion would work ? then we'll take another look on the PR.

Comment (dlind1974) [reply cmd/xl-v1-healing-common.go]: Ok. I have a look at it. Thanks.

**Comment by dlind1974** [comment]
The proposed change by @vadmeste made sense to me. I made updates to the code and the unit test.

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**